source/game/scene/RaceScene.cc

```diff
@@ -23,21 +25,65 @@ RaceScene::~RaceScene() = default;
 
 /// @addr{0x80554208}
 void RaceScene::createEngines() {
-    System::CourseMap::CreateInstance()->init();
-    System::RaceManager::CreateInstance();
-    Field::BoxColManager::CreateInstance();
-    Kart::KartObjectManager::CreateInstance();
-    Field::CollisionDirector::CreateInstance();
-    Item::ItemDirector::CreateInstance();
-    Field::ObjectDirector::CreateInstance();
+    {
+        ScopeLock<GroupID> lock(GroupID::Course);
+        System::CourseMap::CreateInstance()->init();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Race);
+        System::RaceManager::CreateInstance();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Object);
+        Field::BoxColManager::CreateInstance();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Kart);
+        Kart::KartObjectManager::CreateInstance();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Course);
+        Field::CollisionDirector::CreateInstance();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Item);
+        Item::ItemDirector::CreateInstance();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Object);
+        Field::ObjectDirector::CreateInstance();
+    }
 }
 
 /// @addr{0x8055472C}
 void RaceScene::initEngines() {
-    Kart::KartObjectManager::Instance()->init();
-    System::RaceManager::Instance()->init();
-    Item::ItemDirector::Instance()->init();
-    Field::ObjectDirector::Instance()->init();
+    {
+        ScopeLock<GroupID> lock(GroupID::Kart);
+        Kart::KartObjectManager::Instance()->init();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Race);
+        System::RaceManager::Instance()->init();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Item);
+        Item::ItemDirector::Instance()->init();
+    }
+
+    {
+        ScopeLock<GroupID> lock(GroupID::Object);
```

Comment:
I just thought of this - you know how variables with no names are destroyed immediately? We could change `ScopeLock` so that its constructor takes in a GroupID and a lambda such that the heap is set to the provided GroupID, the lambda is run, and the GroupID is reset all in the constructor? Then it is immediately destroyed.

Reply:
As mentioned, I think this pattern is more flexible.